packages/@aws-cdk/aws-servicecatalog/lib/private/validation.ts

```diff
@@ -11,6 +11,29 @@ export class InputValidator {
     }
   }
 
+  /**
+   * Validates string matches the allowed regex pattern.
+   */
+  public static validateRegex(resourceName: string, inputName: string, regexp: RegExp, inputString?: string): void {
+    if (inputString !== undefined && !regexp.test(inputString)) {
+      throw new Error(`Invalid ${inputName} for resource ${resourceName}, must match regex pattern ${regexp}, got: '${this.truncateString(inputString, 100)}'`);
+    }
+  }
+
+  /**
+   * Validates string matches the valid URL regex pattern.
+   */
+  public static validateUrl(resourceName: string, inputName: string, inputString?: string): void {
+    this.validateRegex(resourceName, inputName, /^https?:\/\/.*/, inputString);
+  }
+
+  /**
+  * Validates string matches the valid email regex pattern.
+  */
+  public static validateEmail(resourceName: string, inputName: string, inputString?: string): void {
+    this.validateRegex(resourceName, inputName, /^[\w\d.%+\-]+@[a-z\d.\-]+\.[a-z]{2,4}$/i, inputString);
+  }
+
   private static truncateString(string: string, maxLength: number): string {
```

Comment:
All of these functions must have checks added to them for `Token.isUnresolved()`. See here for details: https://github.com/aws/aws-cdk/pull/15140/files#r654595244.